AWG - RFC

alon opened the pull request:
This is related to issue #2 

It is partially I think correct and partially definitely not.

I would like to get comments on:
enums - is this the right way?
PicoDriver trait extension - ditto? especially the use of unimplemented macro.
ps2000a implementation

Lastly the PicoStreamingDevice additions are pure throw away code, just for testing. What worked was basically:
```rust
    fn connect_to_scope_real(&mut self) {
        // ... get a device the regular way
        let stream_device = device.into_streaming_device();

        //stream_device.set_sig_gen_arbitrary();
        stream_device.set_sig_gen_built_in_v2();

        stream_device.enable_channel(PicoChannel::A, PicoRange::X1_PROBE_2V, PicoCoupling::DC);
        stream_device.enable_channel(PicoChannel::B, PicoRange::X1_PROBE_1V, PicoCoupling::AC);
        // When handler goes out of scope, the subscription is dropped

        stream_device.new_data.subscribe(self.handler.clone());
        stream_device.start(1_000_000).unwrap();

        self.stream_device = Some(stream_device);
    }
```

alon committed:
add call to ps2000aSetSigGenArbitrary via PicoDriver::set_sig_gen_arbitrary

alefminus committed:
now get BUSY

alefminus committed:
add sig_gen_built_in_v2

alon committed:
export software trigger, and built in properties change - missing awg properties change, and some helpers

alon committed:
add wave_type enum

alon committed:
enums: white space fixes, remove duplicate enum name in variants

alon committed:
add some documentation from the sdk

alon committed:
make PicoSigGenTrigSource comparable for egui radio_button

timfish commented:
For the enums variants, I would shorten them like:
```rust
pub enum PicoSigGenTrigType
{
    Rising = 0,
    Falling = 1,
    ...
}
```

For the `PicoStreamingDevice` functions, rather than many arguments, it might be nicer to put them into a struct and then implement `Default`. This way you can leave out properties that aren't important and use it like:

```rust
device.set_sig_gen_properties_built_in(SigGenProperties { 
  trigger_type: PicoTriggerType::Rising,
  ..Default::default()  
});
```
Having not used the sig gen/awg APIs on these devices I don't know how viable default options are.
I still need to find some time to try out these APIs and also see how they compare across the many drivers. 

The higher level APIs .

The `PicoStreamingDevice` attempts to try and hide all the differences between all the drivers/devices and was primarily for my own use case which was streaming raw, gap-less data at high speed from any Pico device. It also automatically reconfigures and restarts streaming if it gets disconnected and stores the device config required to do this. 

What are your plans for siggen/awg while streaming?

alon commented:
I will do both suggestions, they make sense.
Meanwhile I implemented the software trigger function and one of the on the fly update functions since the initial PR.

My aim is to allow software triggering to allow a complete capture. Right now I have some unexpected delays that probably result from misuse of the API.

Doing stop, then calling sig gen, then enable again already works. It is very fast jn human terms, although there must be some lost samples, but I did not measure that yet.

On April 12, 2022 1:32:48 AM GMT+03:00, Tim Fish ***@***.***> wrote:
>For the enums variants, I would shorten them like:
>```rust
>pub enum PicoSigGenTrigType
>{
>    Rising = 0,
>    Falling = 1,
>    ...
>}
>```
>
>For the `PicoStreamingDevice` functions, rather than many arguments, it might be nicer to put them into a struct and then implement `Default`. This way you can leave out properties that aren't important and use it like:
>
>```rust
>device.set_sig_gen_properties_built_in(SigGenProperties { 
>  trigger_type: PicoTriggerType::Rising,
>  ..Default::default()  
>});
>```
>Having not used the sig gen/awg APIs on these devices I don't know how viable default options are.
>I still need to find some time to try out these APIs and also see how they compare across the many drivers. 
>
>The higher level APIs .
>
>The `PicoStreamingDevice` attempts to try and hide all the differences between all the drivers/devices and was primarily for my own use case which was streaming raw, gap-less data at high speed from any Pico device. It also automatically reconfigures and restarts streaming if it gets disconnected and stores the device config required to do this. 
>
>What are your plans for siggen/awg while streaming?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/meatysolutions/pico-sdk/pull/8#issuecomment-1095655100
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

alon committed:
enums: sig_gen related: shorten variants to remove enum prefix

alon committed:
remove streaming wrappers for sig_gen_software_control - confusing, meaning changes per trigger type

alon committed:
introduce SweepShotCount to use an enum for the special 2^32-1 value

alon committed:
streaming/set_sig_gen_built_in_v2: return PicoResult

alon committed:
siggen: better shots/sweeps encoding using a rust enum

alon committed:
prelude: expose PicoResult

alon committed:
add SetSigGenArbitraryProperties

alefminus committed:
add sig_gen_arbitrary_min_max_values

alon committed:
expose sig_gen

alefminus committed:
add phase_props

alefminus committed:
add sig_gen_frequency_to_phase

alefminus committed:
exporet MinMaxArbitraryValues

alefminus committed:
add ps2000a sig_gen_frequency_to_phase

alon committed:
fix typo

alon committed:
add Default trait impl for some enums

alon committed:
move sig-gen types into sig-gen module, add property struct for sig-gen-builtin-v2

alefminus committed:
remove cloning of arbitrary waveform, and a todo

alefminus committed:
ps2000a: slight refactor to show addition for signal generation

alefminus committed:
implement signal generation API for PS4000A (tested on 4824)

alefminus committed:
wip: panic in set_sig_gen_arbitrary if called in open mode - since we will just get PICO_BUSY

wip since docs say in some cases it is acceptable, but I have not seen
it and in practice I just want to change the waveform which is
definitely not possible.

alefminus committed:
streaming/lib: add PartialEq<Target> for State

alefminus committed:
PicoStreamingDevice: add pub wait_for_state_settle: TODO ask author

alefminus committed:
bugfix awg: FOLD earlier: take read lock, not write, on current_state

alefminus committed:
wip add debug print

alefminus committed:
add dwell_count to SigGenArbitraryMinMaxValues

alefminus committed:
formatter churn
